sdk/core/azure-core/src/datetime.cpp

```diff
@@ -4,869 +4,832 @@
-      if ((fracSec != 0 && fractionFormat != TimeFractionFormat::Truncate)
-          || fractionFormat == TimeFractionFormat::AllDigits)
+    auto const fracSecTimeAdjustment
+        = (localDiffHours * OneHourIn100ns) + (localDiffMinutes * OneMinuteIn100ns);
+
+    if (fracSecTimeAdjustment > 0)
+    {
+      if (year == 1 && month == 1)
       {
-        // Append fractional second, which is a 7-digit value with no trailing zeros
-        // This way, '1200' becomes '00012'
-        size_t appended =
-#if defined(_MSC_VER)
-            sprintf_s(
-#else
-            std::sprintf(
-#endif
-                outCursor,
-#if defined(_MSC_VER)
-                9,
-#endif
-                ".%07d",
-                fracSec);
+        auto const fracSecSince0001_01_01 = ((static_cast<int64_t>(day) - 1) * OneDayIn100ns)
+            + (hour * OneHourIn100ns) + (minute * OneMinuteIn100ns) + (second * OneSecondIn100ns)
+            + fracSec + (roundFracSecUp ? 1 : 0);
 
-        if (fractionFormat != TimeFractionFormat::AllDigits)
+        if ((fracSecSince0001_01_01 - fracSecTimeAdjustment) < 0)
         {
-          while (outCursor[appended - 1] == '0')
-          {
-            --appended; // trim trailing zeros
-          }
+          throw std::invalid_argument(
+              "Time zone time adjustments result in a UTC date prior to 0001-01-01.");
         }
-        outCursor += appended;
       }
+    }
+    else if (year == 9999 && month == 12)
+    {
+      auto const fracSecSince9999_12_01 = ((static_cast<int64_t>(day) - 1) * OneDayIn100ns)
+          + (hour * OneHourIn100ns) + (minute * OneMinuteIn100ns) + (second * OneSecondIn100ns)
+          + fracSec + (roundFracSecUp ? 1 : 0);
 
-      *outCursor = 'Z';
-      ++outCursor;
-      return std::string(outBuffer, outCursor);
+      constexpr auto FracSecTo9999_12_31_23_59_59_9999999 = (31 * OneDayIn100ns)
+          + (23 * OneHourIn100ns) + (59 * OneMinuteIn100ns) + (59 * OneSecondIn100ns) + 9999999;
 
-    default:
-      throw std::invalid_argument("Unrecognized date format.");
+      if (fracSecSince9999_12_01 - fracSecTimeAdjustment > FracSecTo9999_12_31_23_59_59_9999999)
+      {
+        throw std::invalid_argument(
+            std::string(
+                fracSecTimeAdjustment < 0
+                    ? "Time zone"
+                    : (second == 60 ? "Leap second" : "Second fraction round up"))
+            + " time adjustments result in a UTC date after 9999-12-31T23:59:59.9999999.");
+      }
+    }
   }
 }
 
-namespace {
-constexpr bool StringStartsWith(char const* haystack, char const* needle)
+static std::string const DayNames[7] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
+static std::string const MonthNames[12]
+    = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
+
+constexpr int32_t DivideAndUpdateRemainder(int64_t* remainder, int64_t value)
+{
+  auto retval = static_cast<int32_t>(*remainder / value);
```

Comment:
Why is this `static_cast` safe to do and not lossy?

Reply:
Purely from usage / points of invocation we know that. This is a helper function that is carefully used within this class.
If there are problems, unit tests will find, by testing broader behavior.